primitives/src/circuit/merkle_tree/mod.rs

```diff
@@ -30,55 +32,226 @@ mod rescue_merkle_tree;
-        merkle_proof: Self::MerklePathVar,
-        merkle_root: Variable,
+        leaf_idx_var: Variable,
+        proof_var: Self::MembershipProofVar,
+        root_var: Variable,
+    ) -> Result<BoolVar, CircuitError>;
+
+    /// Enforce correct `proof_var` for the `leaf_idx_var` against
+    /// `expected_root_var`.
+    fn enforce_membership_proof(
+        &mut self,
+        leaf_idx_var: Variable,
+        proof_var: Self::MembershipProofVar,
+        expected_root_var: Variable,
+    ) -> Result<(), CircuitError>;
+}
+
+/// Gadget for the universal Merkle tree
+///
+/// # Examples
+///
+/// ```
+/// use ark_bls12_377::Fq;
+/// use jf_primitives::circuit::merkle_tree::{MerkleTreeGadget, UniversalMerkleTreeGadget};
+/// use jf_relation::{Circuit, PlonkCircuit};
+/// use jf_primitives::merkle_tree::{MerkleTreeScheme, MerkleCommitment, UniversalMerkleTreeScheme,
+///     prelude::RescueSparseMerkleTree};
+/// use hashbrown::HashMap;
+/// use num_bigint::BigUint;
+///
+/// type SparseMerkleTree<F> = RescueSparseMerkleTree<BigUint, F>;
+/// let mut circuit = PlonkCircuit::<Fq>::new_turbo_plonk();
+/// // Create a 3-ary universal MT, instantiated with a Rescue-based hash, of height 2.
+/// let mut hashmap = HashMap::new();
+/// hashmap.insert(BigUint::from(1u64), Fq::from(2u64));
+/// hashmap.insert(BigUint::from(2u64), Fq::from(2u64));
+/// hashmap.insert(BigUint::from(1u64), Fq::from(3u64));
+/// let mt = SparseMerkleTree::<Fq>::from_kv_set(2, &hashmap).unwrap();
+/// let expected_root = mt.commitment().digest();
+/// // Get a proof for the element in position 2
+/// let proof = mt.universal_lookup(&BigUint::from(3u64)).expect_not_found().unwrap();
+///
+/// // Circuit computation with a MT
+/// let non_leaf_idx_var = circuit.create_variable(BigUint::from(3u64).into()).unwrap();
+///
+/// let proof_var =
+///     UniversalMerkleTreeGadget::<SparseMerkleTree<Fq>>::create_non_membership_proof_variable(
+///         &mut circuit,
+///         &proof
+///     )
+///     .unwrap();
+/// let root_var =
+///     MerkleTreeGadget::<SparseMerkleTree<Fq>>::create_root_variable(
+///         &mut circuit,
+///         expected_root
+///     )
+///     .unwrap();
+/// UniversalMerkleTreeGadget::<SparseMerkleTree<Fq>>::enforce_non_membership_proof(
+///     &mut circuit,
+///     non_leaf_idx_var,
+///     proof_var,
+///     root_var
+/// )
+/// .unwrap();
+/// assert!(circuit.check_circuit_satisfiability(&[]).is_ok());
+/// ```
+pub trait UniversalMerkleTreeGadget<M>: MerkleTreeGadget<M>
+where
+    M: MerkleTreeScheme,
+{
+    /// Type to represent the merkle non-membership proof of the concrete MT
+    /// instantiation. It is MT-specific, e.g arity will affect the exact
+    /// definition of the underlying Merkle path.
+    type NonMembershipProofVar;
+
+    /// Allocate a variable for the membership proof.
+    fn create_non_membership_proof_variable(
+        &mut self,
+        membership_proof: &M::MembershipProof,
```

Comment:
this makes more sense?
```rust 
pub trait UniversalMerkleTreeGadget<M>: MerkleTreeGadget<M>
where
    M: UniversalMerkleTreeScheme,

fn create_non_membership_proof_variable(
        &mut self,
        proof: &M::NonMembershipProof,
)
```

Reply:
Yes, good catch!

Reply:
Done in https://github.com/EspressoSystems/jellyfish/pull/172/commits/f2d2e8e0aa915c3025ac117c040a3b3ff93e2c36